File: BE/dilip/app/services/service_request_service.py
from sqlalchemy.orm import Session
from sqlalchemy import and_, func
from app.models import ServiceRequest
from app.schemas import ServiceRequestCreate, ServiceRequestUpdate
from typing import List, Optional
from datetime import datetime, date

class ServiceRequestService:
    @staticmethod
    async def get_all(db: Session) -> List[ServiceRequest]:
        return db.query(ServiceRequest).all()
    
    @staticmethod
    async def get_by_id(db: Session, request_id: int) -> Optional[ServiceRequest]:
        return db.query(ServiceRequest).filter(ServiceRequest.id == request_id).first()
    
    @staticmethod
    async def get_by_ticket_id(db: Session, ticket_id: str) -> Optional[ServiceRequest]:
        return db.query(ServiceRequest).filter(ServiceRequest.ticket_id == ticket_id).first()
    
    @staticmethod
    async def get_by_client_id(db: Session, client_id: int) -> List[ServiceRequest]:
        return db.query(ServiceRequest).filter(ServiceRequest.client_id == client_id).all()
    
    @staticmethod
    async def get_by_technician_id(db: Session, technician_id: int) -> List[ServiceRequest]:
        return db.query(ServiceRequest).filter(ServiceRequest.assigned_to == technician_id).all()
    
    @staticmethod
    async def get_by_status(db: Session, status: str) -> List[ServiceRequest]:
        return db.query(ServiceRequest).filter(ServiceRequest.status == status).all()
    
    @staticmethod
    async def get_by_priority(db: Session, priority: str) -> List[ServiceRequest]:
        return db.query(ServiceRequest).filter(ServiceRequest.priority == priority).all()
    
    @staticmethod
    async def get_open_tickets(db: Session) -> List[ServiceRequest]:
        return db.query(ServiceRequest).filter(
            ServiceRequest.status.in_(['open', 'assigned', 'in_progress'])
        ).all()
    
    @staticmethod
    async def get_resolved_today(db: Session) -> List[ServiceRequest]:
        today = date.today()
        return db.query(ServiceRequest).filter(
            and_(
                ServiceRequest.status == 'resolved',
                func.date(ServiceRequest.updated_at) == today
            )
        ).all()
    
    @staticmethod
    async def create(db: Session, request_data: ServiceRequestCreate) -> ServiceRequest:
        db_request = ServiceRequest(**request_data.dict())
        db.add(db_request)
        db.commit()
        db.refresh(db_request)
        return db_request
    
    @staticmethod
    async def update(db: Session, request_id: int, request_data: ServiceRequestUpdate) -> Optional[ServiceRequest]:
        db_request = db.query(ServiceRequest).filter(ServiceRequest.id == request_id).first()
        if not db_request:
            return None
        
        update_data = request_data.dict(exclude_unset=True)
        for field, value in update_data.items():
            setattr(db_request, field, value)
        
        # Update the updated_at timestamp
        db_request.updated_at = datetime.utcnow()
        
        db.commit()
        db.refresh(db_request)
        return db_request
    
    @staticmethod
    async def delete(db: Session, request_id: int) -> bool:
        db_request = db.query(ServiceRequest).filter(ServiceRequest.id == request_id).first()
        if not db_request:
            return False
        
        db.delete(db_request)
        db.commit()
        return True 

File: FE/src/services/ServiceRequestService.ts

import { ServiceRequest } from '../types';
import { mockServiceRequests } from '../data/mockData';

export class ServiceRequestService {
  private static requests: ServiceRequest[] = [...mockServiceRequests];

  static async getAll(): Promise<ServiceRequest[]> {
    await new Promise(resolve => setTimeout(resolve, 100));
    return [...this.requests];
  }

  static async getById(id: string): Promise<ServiceRequest | null> {
    await new Promise(resolve => setTimeout(resolve, 50));
    return this.requests.find(request => request.id === id) || null;
  }

  static async getByClientId(clientId: string): Promise<ServiceRequest[]> {
    await new Promise(resolve => setTimeout(resolve, 100));
    return this.requests.filter(request => request.clientId === clientId);
  }

  static async getByTechnicianId(technicianId: string): Promise<ServiceRequest[]> {
    await new Promise(resolve => setTimeout(resolve, 100));
    return this.requests.filter(request => request.assignedTo === technicianId);
  }

  static async create(requestData: Omit<ServiceRequest, 'id' | 'createdAt' | 'updatedAt'>): Promise<ServiceRequest> {
    await new Promise(resolve => setTimeout(resolve, 200));
    const newRequest: ServiceRequest = {
      ...requestData,
      id: Date.now().toString(),
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString()
    };
    this.requests.push(newRequest);
    return newRequest;
  }

  static async update(id: string, updates: Partial<ServiceRequest>): Promise<ServiceRequest | null> {
    await new Promise(resolve => setTimeout(resolve, 200));
    const index = this.requests.findIndex(request => request.id === id);
    if (index === -1) return null;
    
    this.requests[index] = { 
      ...this.requests[index], 
      ...updates, 
      updatedAt: new Date().toISOString() 
    };
    return this.requests[index];
  }

  static async delete(id: string): Promise<boolean> {
    await new Promise(resolve => setTimeout(resolve, 150));
    const index = this.requests.findIndex(request => request.id === id);
    if (index === -1) return false;
    
    this.requests.splice(index, 1);
    return true;
  }
}
